### PR TITLE
Fixed #32095 -- Made QuerySet.update_or_create() save only fields passed in defaults or with custom pre_save().

### PR DESCRIPTION
**Pull request by sarahboyce**
Extended from @apollo13 patch: https://github.com/django/django/pull/13526
Tried to address the PR comments which were to add handling for the auto_now fields and a test for the mti case.

### CODE REVIEW

File: tests/get_or_create/tests.py [hunk 1]
```diff
@@ -503,6 +505,45 @@ def raise_exception():
...
+            self.assertNotIn("name", query)
+
+    def test_mti_update_non_local_concrete_fields(self):
+        journalist = Journalist.objects.create(name="Jane", specialty="Politics")
+        journalist, created = Journalist.objects.update_or_create(
+            pk=journalist.pk,
+            defaults={"name": "John"},
+        )
+        self.assertFalse(created)
+        self.assertEqual(journalist.name, "John")
```
Comment (felixxm) [inline tests/get_or_create/tests.py:545]: `test_mti_update_non_local_concrete_fields` works without the patch, so we should move it to a separate commit.

**Review by felixxm — COMMENTED** [review]
@sarahboyce Thanks for this patch :+1: Some comments from the original patch are not addressed:
- https://github.com/django/django/pull/13526/files#r503304168
- https://github.com/django/django/pull/13526#issuecomment-706707674

Can you take a look?

File: tests/get_or_create/tests.py (hunk 1 above)
Comment (sarahboyce) [reply tests/get_or_create/tests.py]: [8a8ed41](https://github.com/django/django/pull/16072/commits/8a8ed41b87e6031d9a5d28ea82156753a2defd13)

**Comment by sarahboyce** [comment]
> @sarahboyce Thanks for this patch +1 Some comments from the original patch are not addressed:
> 
> * https://github.com/django/django/pull/13526/files#r503304168
> * [#13526 (comment)](https://github.com/django/django/pull/13526#issuecomment-706707674)
> 
> Can you take a look?

Ah I missed these sorry!

Added some changes, I still need to have a think about the wastefulness part of this comment https://github.com/django/django/pull/13526#issuecomment-706707674
> It feels kind of wasteful to compute that on every usage of save(update_fields) twice though.

I'll have a think, also happy for any suggestions :+1:

File: django/db/models/query.py [hunk 2]
```diff
@@ -963,7 +963,26 @@ def update_or_create(self, defaults=None, **kwargs):
...
+
+                for field in self.model._meta.concrete_fields:
+                    # `auto_now` fields need to be within `update_fields` to be updated
+                    if (
+                        getattr(field, "auto_now", False)
+                        and field.name not in update_fields
+                    ):
+                        update_fields.add(field.name)
+                        if field.name != field.attname:
+                            update_fields.add(field.attname)
```
Comment (ngnpope) [inline django/db/models/query.py:981]: Perhaps we could also move this out as a cached property alongside `_concrete_field_names`?

We'd probably want to name it in a way that can be used for any field that updates on save in case anything is added in the future, e.g. `_auto_update_field_names`.

Then this becomes:
```python
            update_fields = set(defaults)
            concrete_field_names = self.model._meta._concrete_field_names

            # `update_fields` does not support non-concrete fields (#31382)
            if concrete_field_names.issuperset(update_fields):
                # Add fields that are automatically updated, e.g. `auto_now` fields.
                # This is to maintain backward compatibility as these fields are not
                # updated if not specified explicitly in the `update_fields` list.
                update_fields |= self.model._meta._auto_update_field_names
                obj.save(using=self.db, update_fields=update_fields)

            else:
                obj.save(using=self.db)
```

Note that I've also updated the comment. We should be clear that, without this, things will break. Semi-related, I've [mentioned before](https://github.com/django/django/pull/13526#pullrequestreview-506449403) that it'd be nice to add some documentation that makes it clear that `update_fields` without specifying the fields with `add_now` won't update them.

Comment (sarahboyce) [reply django/db/models/query.py]: Pushed up a change thank you, looks cleaner :+1: 


Good point about adding to the docs, was looking [here](https://docs.djangoproject.com/en/4.1/ref/models/instances/#what-happens-when-you-save) where it says what happens when you save, point 2 says: 

> **Preprocess the data.** Each field’s [pre_save()](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.Field.pre_save) method is called to perform any automated data modification that’s needed. For example, the date/time fields override pre_save() to implement [auto_now_add](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.DateField.auto_now_add) and [auto_now](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.DateField.auto_now).

Which is maybe misleading in the `update_fields` case.
I can try to add a note in [here](https://docs.djangoproject.com/en/4.1/ref/models/instances/#what-happens-when-you-save) that this pre-processing step is skipped for those fields unless explicitly added into the `update_fields`.
Might not be the best place to put it/word it but I can give it a shot. @ngnpope what do you think?

Also should I put that in a separate PR or do you think it should be included in this one? :thinking:

**Comment by apollo13** [comment]
Thank you for picking up my old PR :)

File: django/db/models/query.py (hunk 2 above)
Comment (sarahboyce) [reply django/db/models/query.py]: [e542e6f](https://github.com/django/django/pull/16072/commits/e542e6f12c7be90d17cf30258e362c33804d488c) I gave it a  go, let me know if it reads ok :+1:

File: tests/get_or_create/tests.py [hunk 3]
```diff
@@ -503,6 +505,45 @@ def raise_exception():
         )
         self.assertFalse(created)
 
+    def test_mti_update_non_local_concrete_fields(self):
+        journalist = Journalist.objects.create(name="Jane", specialty="Politics")
+        journalist, created = Journalist.objects.update_or_create(
+            pk=journalist.pk,
+            defaults={"name": "John"},
+        )
+        self.assertFalse(created)
```
Comment (David-Wobrock) [inline tests/get_or_create/tests.py:514]: Use `assertIs` here and below instead of `assertFalse`:
```suggestion
        self.assertIs(created, False)
```
(see [Coding style docs](https://docs.djangoproject.com/en/4.1/internals/contributing/writing-code/coding-style/#python-style))

File: django/db/models/options.py [hunk 4]
```diff
@@ -982,6 +982,30 @@ def _property_names(self):
...
+                if field.name != field.attname:
+                    names.append(field.attname)
+        return frozenset(names)
+
+    @cached_property
+    def _auto_update_field_names(self):
+        """Return a set of field names which auto update on model save."""
+        names = []
+        for field in self.concrete_fields:
+            if getattr(field, "auto_now", False):
```
Comment (David-Wobrock) [inline django/db/models/options.py:1002]: Is `auto_now` the only case where we have this behaviour?
Would there be a way to detect if a user defined a custom model field that would also have a similar behaviour, but doesn't have an `auto_now`?
Looking at some implementations in Django, `django.db.models.fields.files.FileField` is by default saving the file if not committed yet. We perhaps need to handle those cases too? 🤔

**Review by David-Wobrock — COMMENTED** [review]
Cool to see this being worked on 💪

File: tests/get_or_create/tests.py (hunk 3 above)
Comment (sarahboyce) [reply tests/get_or_create/tests.py]: Ah good spot thank you, updated :+1:

File: django/db/models/options.py (hunk 4 above)
Comment (sarahboyce) [reply django/db/models/options.py]: You're completely right, I have made some edits to move it away from specifically checking for auto_now fields and to instead add any field which is edited on the fields pre_save (which includes the auto_now fields)
Let me know what you think :+1:

File: docs/ref/models/instances.txt [hunk 5]
```diff
@@ -621,6 +621,13 @@ from the DB will get updated. In effect there is an automatic
 ``update_fields`` in this case. If you assign or change any deferred field
 value, the field will be added to the updated fields.
 
+.. note::
+    If ``update_fields`` is specified, only the
+    :meth:`~django.db.models.Field.pre_save` methods of the ``update_fields``
+    will be called. For example, this means that date/time fields with
+    ``auto_now=True`` will not be updated unless they are included in the
+    ``update_fields``.
```
[the remark this answers is not in the document]
Comment (sarahboyce) [reply docs/ref/models/instances.txt:629]: Thank you for the review @David-Wobrock :star: 

I think @ngnpope has explained it quite well [here](https://github.com/django/django/pull/13526#pullrequestreview-506449403) about the `update_fields` behaviour.

Essentially, what I have added to the docs is clarifying the **existing** behaviour of `update_fields` and so doesn't need a release note/is not a breaking change.

Then the change we are adding is to use `update_fields` under the hood of `update_or_create` (as a potential performance optimisation).
As before it was doing a standard save, the pre_save fields were being handled.
So we are adding some logic to not change the current behaviour of `update_or_create` as a unintended consequence of this optimisation change.

I have made a slight update to the comment, let me know if it is clearer and also feel free to suggest any further edits!

**Comment by sarahboyce** [comment]
I realise I will probably not be popular for adding freeze_gun into the testing requirements.
I was having that all tests except the windows runner tests were passing and I've had refactor the tests a bit to have the full suite pass, let me know if anyone has any suggestions :+1:

File: tests/get_or_create/tests.py [hunk 6]
```diff
@@ -503,6 +507,45 @@ def raise_exception():
...
+        book = Book.objects.create(publisher=publisher, name="The Book of Ed & Fred")
+
+        for defaults, update_time in [
+            ({"publisher": publisher}, datetime(2013, 3, 3, 3, 33, 33, 333333)),
+            ({"publisher_id": publisher}, datetime(2014, 4, 4, 4, 44, 44, 444444)),
+        ]:
+            with self.subTest(defaults=defaults, update_time=update_time):
+
+                with CaptureQueriesContext(connection) as captured_queries:
+                    with freezegun.freeze_time(update_time):
```
Comment (David-Wobrock) [inline tests/get_or_create/tests.py:530]: Would `django.test.utils.freeze_time` be sufficient here?
And therefore we don't need to add `freezegun` 🤞

File: docs/ref/models/instances.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (David-Wobrock) [reply docs/ref/models/instances.txt:629]: Ah yes, of course 🤦  My bad!  🙌 

This note is a great addition IMHO 👍

File: django/db/models/query.py [hunk 7]
```diff
@@ -963,7 +963,25 @@ def update_or_create(self, defaults=None, **kwargs):
...
+            concrete_field_names = self.model._meta._non_pk_concrete_field_names
+
+            # `update_fields` does not support non-concrete fields (#31382)
+            if concrete_field_names.issuperset(update_fields):
+                # Add fields which are set on pre_save, e.g. `auto_now` fields.
+                # This is to maintain backward compatibility as these fields are not
+                # updated if not specified explicitly in the `update_fields` list.
+                for field in self.model._meta.local_concrete_fields:
+                    if not field.primary_key:
+                        if getattr(obj, field.attname) != field.pre_save(obj, False):
```
Comment (David-Wobrock) [inline django/db/models/query.py:977]: Arf, this is also not optimal either.

`pre_save` can have side-effects, like `django.db.models.fields.files.FileField.pre_save` does 😕 
We probably don't want to trigger those here.

I mean, serendipitously it would work for the `FileField` because even if the returned value is still the same (so we don't add the `field.name` to `updated_fields`), we actually triggered the side-effect committing the file 😂 
However, that seems pretty brittle 😅 

I'm not sure what the cleanest/Djangoest approach would be here 🤔 
We could add an attribute on the Field class, like `Field.has_pre_save: bool`, but that creates a precedent and users/libs must update their code accordingly.
But at least, we would know _for sure_ which fields need to be added and which don't.

Any other suggestion is very welcome!

Comment (sarahboyce) [reply django/db/models/query.py]: I am now comparing the bytecode of the functions which allows us to see if it has a custom `pre_save` without having to call the `pre_save`'s themselves.

I have never used this before, it seems to work. Might be a little unusual However the tests are passing and it doesn't require any one to update their code. :shrug: 

The `FileField` will be included in the update_fields but I guess that is ok, as before we were calling the pre_save on all fields as they were using the standard save - so it is kind of existing behaviour.

File: tests/get_or_create/tests.py [hunk 8]
```diff
@@ -503,6 +505,39 @@ def raise_exception():
...
+    def test_update_only_defaults_and_pre_save_fields_when_local_fields(self):
+        publisher = Publisher.objects.create(name="Acme Publishing")
+        book = Book.objects.create(publisher=publisher, name="The Book of Ed & Fred")
+
+        for defaults in [{"publisher": publisher}, {"publisher_id": publisher}]:
+            with self.subTest(defaults=defaults):
+                with CaptureQueriesContext(connection) as captured_queries:
+                    book, created = Book.objects.update_or_create(
+                        pk=book.pk,
+                        defaults={"publisher": publisher},
```
Comment (felixxm) [inline tests/get_or_create/tests.py:526]: :thinking:
```suggestion
                        defaults=defaults,
```

File: django/db/models/query.py (hunk 7 above)
Comment (kezabelle) [reply django/db/models/query.py]: My recollection is a touch fuzzy/uncertain, but you _may_ be able to get away with _not_ looking at at the deep internals (`co_code`), and just comparing `Field.pre_save is field_instance.__class__.pre_save`
Though to caveat that, there may be circumstances (hello multiprocesses?) where that doesn't work?

File: tests/get_or_create/tests.py (hunk 8 above)
Comment (sarahboyce) [reply tests/get_or_create/tests.py]: :facepalm: thank you

File: django/db/models/query.py (hunk 7 above)
Comment (sarahboyce) [reply django/db/models/query.py]: @kezabelle that's so much nicer thank you :star:

File: django/db/models/query.py [hunk 9]
```diff
@@ -963,7 +963,27 @@ def update_or_create(self, defaults=None, **kwargs):
...
+            # `update_fields` does not support non-concrete fields (#31382)
+            if concrete_field_names.issuperset(update_fields):
+                # Add fields which are set on pre_save, e.g. `auto_now` fields.
+                # This is to maintain backward compatibility as these fields are not
+                # updated if not specified explicitly in the `update_fields` list.
+                for field in self.model._meta.local_concrete_fields:
+                    if (
+                        not field.primary_key
+                        and field.__class__.pre_save is not Field.pre_save
+                    ):
```
Comment (felixxm) [inline django/db/models/query.py:979]: Maybe :thinking: 
```suggestion
                    if not (
                        field.primary_key
                        or field.__class__.pre_save is Field.pre_save
                    ):
```
according to de Morgan's laws.

**Review by felixxm — COMMENTED** [review]
@sarahboyce Many thanks :+1: I will start pushing cleanups to separate PRs (docs clarification, extra tests, etc.). We can focus on the main change after they are merged.

**Review by David-Wobrock — APPROVED** [review]

File: django/db/models/query.py (hunk 9 above)
Comment (sarahboyce) [reply django/db/models/query.py]: Updated thank you :+1: also rebased so it is just the 1 commit

**Review by felixxm — APPROVED** [review]
@sarahboyce Thanks for all your efforts :star2: 

I pushed tiny final edits.

**Comment by felixxm** [comment]
buildbot, test on oracle.
